### PR TITLE
FIPS: audit and switch some memset & bzero users

### PR DESCRIPTION
**Pull request by rzarzynski**
For the sake of compliance with FIPS memory where security material (like keys) was stored, should be cleaned when it isn't used anymore. This is intended to limit the impact of other security problems allowing to inspect memory.

In many cases such sanitization is performed with `memset` or `bzero` to zeroize the memory. However, C++ language, due to [the as-if rule](https://en.cppreference.com/w/cpp/language/as_if), provides less guarantees than necessary to ensure that a call to e.g. `memset` will be always translated into intended stores. This isn't something specific to it nor `bzero`. All a compiler needs to know to perform dead store elimination is the code itself (to exclude e.g. `volatile` fencing). Presumably it could assume how the things from standard library look internally – [like with an inlineable function](https://godbolt.org/z/XJnAnA).

The problem was already discussed in the GCC bug reports which finally have been marked as "invalid":
  * https://gcc.gnu.org/bugzilla/show_bug.cgi?id=8537,
  * https://gcc.gnu.org/bugzilla/show_bug.cgi?id=71388.

Because of that we want to perform the security clean-ups with a dedicated procedure that takes responsibility of prohibiting compilers from optimizing it out. OpenSSL already provides such utility: `OPESNSSL_cleanse`. The first commit integrates it into Ceph's abstractions over crypto.
Its intended clients are some current users of `memset` and `bzero` found using the GCC's `deprecated` attribute. See: https://gist.github.com/rzarzynski/db9b4ca6b3d409d2ab8d38f4d6678063.

A subset of the detected users are audited and switched to `ceph::crypto::zeroize_for_security` by further commits.

Signed-off-by: Radoslaw Zarzynski <rzarzyns@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by mattbenjamin** [comment]
Very detailed.  I like the indirection via ceph::crypto.

Matt

**Comment by mdw-at-linuxbox** [comment]
Ah.  One thing I missed.  Do we really need to comment "this memset is not security related." ?  Also true for init'ing hmac ctx, which I'll go find and comment inline.

File: src/common/ceph_crypto.h [hunk 1]
```diff
@@ -98,7 +98,9 @@ namespace ceph::crypto::ssl {
   public:
     HMAC (const EVP_MD *type, const unsigned char *key, size_t length)
       : mpType(type) {
-      ::memset(&mContext, 0, sizeof(mContext));
+      // the strict FIPS zeroization doesn't seem to be necessary here.
+      // just in the case.
+      ::ceph::crypto::zeroize_for_security(&mContext, sizeof(mContext));
```
Comment (mdw-at-linuxbox) [inline src/common/ceph_crypto.h]: Yes, this one really shouldn't be zeroize_for_security.  That context hasn't yet got any nasty secrets in it, so no need to sanitize.

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2019-11-19_09:56:38-rados-wip-kefu2-testing-2019-11-19-1509-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2019-11-21_08:41:20-rados-wip-kefu2-testing-2019-11-19-1509-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2019-11-23_01:15:32-rados-wip-kefu2-testing-2019-11-23-0140-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2019-11-24_13:33:51-rados-wip-kefu2-testing-2019-11-24-1912-distro-basic-mira/

**Pull request merged**